### PR TITLE
[doc] Refactor kernels and functions

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

File: docs/lang/articles/kernels/syntax.md [hunk 1]
```diff
@@ -4,109 +4,120 @@ sidebar_position: 1
...
-Taichi is a domain-specific language embedded in Python. The meaning of "embedded in Python" is twofold:  The code you've written in Taichi is valid Python code but is compiled and executed in Taichi's runtime; the rest of your code is treated as native Python code and executed in Python's virtual machine.
+Taichi is embedded in Python, the code you have written for Taichi look like valid Python code, hence we need a way to differentiate the code for Taichi from the code for Python. We use the two decorators `@ti.kernel` and `@ti.func` for this:
 
-To differentiate the code for Taichi from the code for Python, we use the two decorators `@ti.kernel` and `@ti.func`:
++ Functions decorated with `@ti.kernel` are called Taichi kernels (called kernel for short).
++ Functions decorated with `@ti.func` are called Taichi functions.
 
-- Functions decorated with `@ti.kernel` are called kernels.
-- Functions decorated with `@ti.func` are called Taichi functions.
+Let's see an example:
```
Comment (yuanming-hu) [inline docs/lang/articles/kernels/syntax.md:12]:
```suggestion
You can call Taichi kernel directly from Python code. You can also call Taichi functions from Taichi kernels, but not directly from Python code.
Let's see an example:
```

File: docs/lang/articles/kernels/syntax.md [hunk 2]
```diff
@@ -4,109 +4,120 @@ sidebar_position: 1
...
+ti.init(arch=ti.cpu)
+
+@ti.func
+def inv_square(x):
+    return 1.0 / (x * x)
+
+print(inv_square(1.0))  # syntax error!
+```
+
+Here by "out of the Taichi scope" we mean the scope that is not inside a kernel or a Taichi function. We shall also use the term *Python scope* for short of this.
```
Comment (yuanming-hu) [inline docs/lang/articles/kernels/syntax.md:47]: Let's highlight the definition of Taichi/Python scope :-)

**Review by yuanming-hu — APPROVED** [review]
Cool! Details need more work, but the content LGTM in general!

File: docs/lang/articles/kernels/syntax.md [hunk 3]
```diff
@@ -215,129 +221,51 @@ kernel_2()  # 2
...
-def my_kernel():
-    x = 24
-    my_func(x)
-    print(x)  #  will print 25
-```
-</details>
 
 ### Return values
 
 The return values of a Taichi function can be scalar, `ti.Matrix`, `ti.Vector`, `ti.Struct`, and more. Note that:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:262]:
```suggestion
The return values of a Taichi function can be scalars, `ti.Matrix`, `ti.Vector`, `ti.Struct`, or others. Note that:
```

File: docs/lang/articles/kernels/syntax.md [hunk 4]
```diff
@@ -4,109 +4,134 @@ sidebar_position: 1
...
-Code outside of the Taichi scope is in the Python scope. The code in the Python scope is native Python and executed by Python's virtual machine, *not* by Taichi's runtime.
+```python
+import taichi as ti
+ti.init(arch=ti.cpu)
+
+@ti.func
+def inv_square(x):
+    return 1.0 / (x * x)
+
+print(inv_square(1.0))  # syntax error!
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:44]:
```suggestion
print(inv_square(1.0))  # Syntax error!
```

File: docs/lang/articles/kernels/syntax.md [hunk 5]
```diff
@@ -4,109 +4,134 @@ sidebar_position: 1
...
+
+Here by "out of the Taichi scope" we mean the scope that is not inside a kernel or a Taichi function. We shall also use the term *Python scope* for short of this.
+
+:::tip IMPORTANT
+
+We give the following definitions:
+
+1. The code inside a kernel or a Taichi function is in the **Taichi scope**. The code in the Taichi scope is compiled by Taichi's runtime and executed in parallel on CPU or GPU devices for high-performance computation.
+
+   Taichi scope corresponds to the device side in CUDA.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:55]:
```suggestion
   The Taichi scope corresponds to the device side in CUDA.
```

File: docs/lang/articles/kernels/syntax.md [hunk 6]
```diff
@@ -4,109 +4,134 @@ sidebar_position: 1
...
+
+2. Code outside of the Taichi scope is in the **Python scope**. The code in the Python scope is native Python and executed by Python's virtual machine, not by Taichi's runtime.
+
+   Python scope corresponds to the host side in CUDA.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:59]:
```suggestion
   The Python scope corresponds to the host side in CUDA.
```

File: docs/lang/articles/kernels/syntax.md [hunk 7]
```diff
@@ -4,109 +4,134 @@ sidebar_position: 1
...
-Python scope corresponds to the *host side* in CUDA.
 :::
 
+As you have seen, there are a few differences between kernels and Taichi functions. We now explain their differences and the roles they play in a program and in more detail.
+
+
 ## Kernel
 
-A kernel is the entry point from which Taichi's runtime takes control and the smallest unit for runtime execution. You can define multiple kernels in your program, and kernels are *independent* from each other. You call a kernel the same way you call a Python function, and you are allowed to switch back and forth between Taichi's runtime and Python's virtual machine.
+A kernel is the entry point from which Taichi's runtime takes control and the smallest unit for runtime execution. You can define multiple kernels in your program, and these kernels are *independent* from each other. You call a kernel the same way you call a Python function, and you are allowed to switch back and forth between Taichi's runtime and Python's virtual machine.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:68]:
```suggestion
As the smallest execution unit in Taichi, a kernel is the entry point from which Taichi's runtime takes control. You call a kernel the same way you call a Python function and can switch back and forth between Taichi's runtime and Python's virtual machine.
```

File: docs/lang/articles/kernels/syntax.md [hunk 8]
```diff
@@ -4,109 +4,134 @@ sidebar_position: 1
...
-Taichi's runtime compiles and executes kernels in the order you call them. It stores compiled kernels in the cache so that the next call to the same kernel does not need to be compiled again.
+When there are more than one kernel in a program, Taichi will compile and execute them in the order they are called. The compiled kernels are stored in the cache so that subsequent calls won't invoke re-compilations.
 
 :::caution WARNING
-You must *not* call a kernel from inside another kernel or from inside a Taichi function. You can only call a kernel directly or from inside a native Python function. To put it differently, you can only call a kernel from the *Python scope*.
-:::
 
-:::note
-A kernel corresponds to the `__global__` function in CUDA.
+You must *not* call a kernel from inside another kernel or from inside a Taichi function. You can only call a kernel directly or from inside a native Python function.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:84]:
```suggestion
You must *not* call a kernel from inside another kernel or from inside a Taichi function. You can only call a kernel directly or from inside a native Python function. In other words, you can only call a kernel from inside the Python scope.
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/kernels/syntax.md [hunk 9]
```diff
@@ -4,109 +4,135 @@ sidebar_position: 1
...
+ti.init(arch=ti.cpu)
+
+@ti.func
+def inv_square(x):
+    return 1.0 / (x * x)
+
+print(inv_square(1.0))  # Syntax error!
+```
+
+Here by "out of the Taichi scope" we mean the scope that is not inside a kernel or a Taichi function. We shall also use the term *Python scope* for short of this.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/kernels/syntax.md:48]:
```suggestion
The Taichi function should have fallen in the Taichi scope, a concept as opposed to the "Python scope". 
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

**Review by Olinaaaloompa — APPROVED** [review]

**Pull request merged**